Apps/Src/ReadCarCAN.c

```diff
@@ -68,110 +75,61 @@ static void updateSaturation(int8_t chargeMessage){
-    assertExceptionError(readBPSError);
+    if(chargeMessage == -1){
+        chargeEnable = false;
+    }else if(chargeMsgSaturation >= SATURATION_THRESHOLD){
+        chargeEnable = true;
+    }
 }
 
-// helper function to call if charging should be enabled
-static inline void chargingEnable(void) {
-    OS_ERR err;
-    CPU_TS ts;
-
-    // mark regen as enabled
-    chargeEnable = true;
-
-    // check if we need to run the precharge sequence to turn on the array
-    bool shouldRestartArray = false;
-
-    OSMutexPend(&arrayRestartMutex, 0, OS_OPT_PEND_BLOCKING, &ts, &err);
-    assertOSError(OS_READ_CAN_LOC,err);
-
-    // if the array is off and we're not already turning it on, start turning it on
-    shouldRestartArray = !restartingArray && (Contactors_Get(ARRAY_CONTACTOR)==OFF);
-
-    // wait for precharge for array 
-    if(shouldRestartArray){
-
-            restartingArray = true;
-
-            // Wait to make sure precharge is finished and then restart array
-            OSTmrStart(&prechargeDlyTimer, &err);
-            assertOSError(OS_READ_CAN_LOC, err);
-            
-        }
-    
-
-    OSMutexPost(&arrayRestartMutex, OS_OPT_NONE, &err);
-    assertOSError(OS_READ_CAN_LOC,err);
+/**
+ * @brief Nested function as the same function needs to be executed however the timer requires different parameters
+ * @param p_tmr pointer to the timer that calls this function, passed by timer
+ * @param p_arg pointer to the argument passed by timer
+*/
+static void assertMissedBPSMsg(void *p_tmr, void *p_arg){
+    assertReadCarCANError(READCARCAN_ERR_MISSED_MSG);
 }
 
 /**
  * @brief Callback function for the precharge delay timer. Waits for precharge and then restarts the array.
  * @param p_tmr pointer to the timer that calls this function, passed by timer
  * @param p_arg pointer to the argument passed by timer
- * 
 */
 static void arrayRestart(void *p_tmr, void *p_arg){
-    Minion_Error_t Merr;
     if(chargeEnable){    // If regen has been disabled during precharge, we don't want to turn on the main contactor immediately after
-        Contactors_Set(ARRAY_CONTACTOR, (Minion_Read_Pin(IGN_1, &Merr)), false); //turn on array contactor if the ign switch lets us
-        UpdateDisplay_SetArray(true);
-    }
-    // done restarting the array
-    restartingArray = false;
+        Contactors_Set(ARRAY_CONTACTOR, arrayIgnitionStatus, false); // Turn on array contactor if the ign switch lets us
 
-};
-
-/**
- * @brief Disables charging when canWatchTimer hits 0 (when charge enable messages are missed)
- * @param p_tmr pointer to the timer that calls this function, passed by timer
- * @param p_arg pointer to the argument passed by timer
-*/
-void canWatchTimerCallback (void *p_tmr, void *p_arg){
-    chargingDisable();
-}
+        // Update display based on if array contactor is on/off
+        if(Contactors_Get(ARRAY_CONTACTOR) == ON){
+            UpdateDisplay_SetArray(true);
+        }else{
+            UpdateDisplay_SetArray(false);
```

Comment:
Just curious, what is the need for having them in one place? @IshDeshpa

Reply:
It's far more confusing for a new person to read how the contactors are updated if one of the sets is in a callback for a timer delay and the other is in the main while loop, plus it could lead to some synchronization issues (what if the precharge timer ends, we decide to turn on the contactor, and then we decide to turn it off in the next iteration of the while loop?), plus it's easier to debug and see if the Contactors_Set call is in one spot with some boolean condition determining whether to turn it off/on.